OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -584,50 +584,52 @@ ________________________________________________________________________
 
 double SmoothSegmentedFunction::calcDerivative(double x, int order) const
 {
-    {
+    const double x0 = _smoothData->_x0;
+    const double x1 = _smoothData->_x1;
+
+    if (x < x0) {
         const double y0 = _smoothData->_y0;
-        const double x0 = _smoothData->_x0;
         const double dydx0 = _smoothData->_dydx0;
-        if (x < x0) {
-            switch (order) {
-                case 0: return y0 + dydx0*(x-x0);
-                case 1: return dydx0;
-                default: return 0.;
-            }
+        switch (order) {
+            case 0: return y0 + dydx0*(x-x0);
+            case 1: return dydx0;
+            default: return 0.;
         }
     }
 
-    {
-        const double x1 = _smoothData->_x1;
+    if (x <= x1) {
+        const SimTK::Array_<SimTK::Vec6>& ctrlPtsX = _smoothData->_ctrlPtsX;
+        const int idx  = SegmentedQuinticBezierToolkit::calcIndex(x,ctrlPtsX);
+
+        const SimTK::Array_<SimTK::Spline>& arraySplineUX =
+            _smoothData->_arraySplineUX;
+        const double u = SegmentedQuinticBezierToolkit::calcU(
+                x,
+                ctrlPtsX[idx],
+                arraySplineUX[idx],
+                UTOL,
+                MAXITER);
+
+        const SimTK::Array_<SimTK::Vec6>& ctrlPtsY = _smoothData->_ctrlPtsY;
+        return SegmentedQuinticBezierToolkit::calcQuinticBezierCurveDerivDYDX(
+                u,
+                ctrlPtsX[idx],
+                ctrlPtsY[idx],
+                order);
+    }
+
+    if (x > x1) {
         const double y1 = _smoothData->_y1;
         const double dydx1 = _smoothData->_dydx1;
-        if (x > x1) {
-            switch (order) {
-                case 0: return y1 + dydx1*(x-x1);
-                case 1: return dydx1;
-                default: return 0.;
-            }
+        switch (order) {
+            case 0: return y1 + dydx1*(x-x1);
+            case 1: return dydx1;
+            default: return 0.;
         }
     }
 
-    const SimTK::Array_<SimTK::Vec6>& ctrlPtsX = _smoothData->_ctrlPtsX;
-    const int idx  = SegmentedQuinticBezierToolkit::calcIndex(x,ctrlPtsX);
-
-    const SimTK::Array_<SimTK::Spline>& arraySplineUX =
-        _smoothData->_arraySplineUX;
-    const double u = SegmentedQuinticBezierToolkit::calcU(
-            x,
-            ctrlPtsX[idx],
-            arraySplineUX[idx],
-            UTOL,
-            MAXITER);
-
-    const SimTK::Array_<SimTK::Vec6>& ctrlPtsY = _smoothData->_ctrlPtsY;
-    return SegmentedQuinticBezierToolkit::calcQuinticBezierCurveDerivDYDX(
-            u,
-            ctrlPtsX[idx],
-            ctrlPtsY[idx],
-            order);
+    // In case of NaN return zero.
+    return 0.;
```

Comment:
Do you think this might be hiding another issue? Is it possible that your CMC simulation/visualization is feeding this function a NaN value for `x` which then gets masked by returning zero here. Or do you think the changes introduced new NaNs somewhere?

Reply:
I see, this was the previous behavior of the function. In that case, this is probably fine.